fastly/resource_fastly_custom_dashboard.go

```diff
@@ -0,0 +1,458 @@
+				},
+			},
+		},
+	}
+
+	schemaVisualization = schema.Schema{
+		Type:        schema.TypeList,
+		Required:    true,
+		Description: "An object which describes the data visualization to display.",
+		MaxItems:    1,
+		MinItems:    1,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"config": &schemaVisualizationConfig,
+				"type": {
+					Type:        schema.TypeString,
+					Required:    true,
+					Description: "The type of visualization to display. One of: `chart`.",
+					ValidateDiagFunc: validation.ToDiagFunc(validation.StringInSlice(
+						[]string{"chart"},
+						false,
+					)),
+				},
+			},
+		},
+	}
+
+	schemaVisualizationConfig = schema.Schema{
+		Type:        schema.TypeList,
+		Required:    true,
+		Description: "Configuration options for the selected data source.",
+		MaxItems:    1,
+		MinItems:    1,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"calculation_method": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					Description: "The aggregation function to apply to the dataset. One of: `avg`, `sum`, `min`, `max`, `latest`, `p95`.",
+					ValidateDiagFunc: validation.ToDiagFunc(validation.StringInSlice(
+						[]string{"avg", "sum", "min", "max", "latest", "p95"},
+						false,
+					)),
+				},
+				"format": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					Description: "The units to use to format the data. One of: `number`, `bytes`, `percent`, `requests`, `responses`, `seconds`, `milliseconds`, `ratio`, `bitrate`.",
+					ValidateDiagFunc: validation.ToDiagFunc(validation.StringInSlice(
+						[]string{"number", "bytes", "percent", "requests", "responses", "seconds", "milliseconds", "ratio", "bitrate"},
+						false,
+					)),
+				},
+				"plot_type": {
+					Type:        schema.TypeString,
+					Required:    true,
+					Description: "The type of chart to display. One of: `line`, `bar`, `single-metric`, `donut`.",
+					ValidateDiagFunc: validation.ToDiagFunc(validation.StringInSlice(
+						[]string{"line", "bar", "single-metric", "donut"},
+						false,
+					)),
+				},
+			},
+		},
+	}
+)
+
+func resourceFastlyCustomDashboard() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceFastlyCustomDashboardCreate,
+		ReadContext:   resourceFastlyCustomDashboardRead,
+		UpdateContext: resourceFastlyCustomDashboardUpdate,
+		DeleteContext: resourceFastlyCustomDashboardDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"dashboard_item": {
+				Type:        schema.TypeList,
+				Optional:    true,
+				Description: "A list of dashboard items.",
+				MinItems:    0,
+				MaxItems:    100,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"data_source": &schemaDataSource,
+						"id": {
+							Type:        schema.TypeString,
+							Computed:    true,
```

Comment:
I believe (though still TBD, so just putting a comment) we're going to change this field to be required rather than computed, to resolve the issue of links to charts breaking if a dashboard item is inserted or removed, or if they're reordered.